apps/client/src/flogo/app/core/apps.service.ts

```diff
@@ -139,6 +183,49 @@ export class AppDetailService {
     );
   }
 
+  public hasTriggers() {
+    return this.triggersState$.pipe(
+      take(1),
+      map(triggers => {
+        const appHasTriggers = triggers && triggers.length > 0;
+        return !!appHasTriggers;
+      })
+    );
+  }
+
+  public getAvailableShimBuildOptions$(): Observable<ShimBuildOptions[]> {
+    const triggersUsed$: Observable<Set<string>> = this.groupsByTrigger$.pipe(
+      map(flowGroups => {
+        return new Set(
```

Comment:
can also use Set's chainable interface directly and avoid the creation of the array:

```ts
 return flowGroups.reduce(
    (refsSet, group) =>
      group.trigger ? refsSet.add(group.trigger.ref) : refsSet,
    new Set()
  );
```